api/v1alpha1/ingressnodefirewall_webhook.go

```diff
@@ -43,22 +53,175 @@ var _ webhook.Validator = &IngressNodeFirewall{}
+		allErrs = append(allErrs, err)
+	}
+	if !orderIsUnique(rules) {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("spec").Child("ingress").Index(infRulesIndex).Key("rules"),
+			infName, "must have unique order"))
+	}
+	for ruleIndex, rule := range rules {
+		if err := validateRule(rule, infRulesIndex, ruleIndex, infName); err != nil {
+			allErrs = append(allErrs, err)
+		}
+	}
+	return allErrs
+}
+
+func validateRule(rule IngressNodeFirewallProtocolRule, infRulesIndex, ruleIndex int, infName string) *field.Error {
+	if rule.Protocol == ProtocolTypeICMP {
+		if isValid, reason := isValidICMPRule(rule); !isValid {
+			return field.Invalid(field.NewPath("spec").Child("ingress").Index(infRulesIndex).Key("rules").Index(ruleIndex),
+				infName, fmt.Sprintf("must be a valid ICMP(V6) rule: %s", reason))
+		}
+	}
+
+	if rule.Protocol == ProtocolTypeTCP || rule.Protocol == ProtocolTypeUDP || rule.Protocol == ProtocolTypeSCTP {
+		if isValid, reason := isValidTCPUDPSCTPRule(rule); !isValid {
+			return field.Invalid(field.NewPath("spec").Child("ingress").Index(infRulesIndex).Key("rules").Index(ruleIndex),
+				infName, fmt.Sprintf("must be a valid %s rule: %s", rule.Protocol, reason))
+		}
+
+		if isConflict, reason := isConflictWithPinHoles(rule); isConflict {
+			return field.Forbidden(field.NewPath("spec").Child("ingress").Index(infRulesIndex).Key("rules").Index(ruleIndex),
+				fmt.Sprintf("%s", reason))
+		}
+	}
+	return nil
+}
+
+func isConflictWithPinHoles(rule IngressNodeFirewallProtocolRule) (bool, string) {
+	if conflictSSH(rule) {
+		return true, "must not block TCP port 22 because it is used to login to nodes"
+	} else if conflictDHCP(rule) {
+		return true, "must not block UDP port 68 because it is used to access DHCP"
+	} else if conflictETCD(rule) {
+		return true, "must not block TCP port 2380 or 2379 because it is used to access ETCD"
+	} else if conflictAPIServer(rule) {
+		return true, "must not block TCP port 6443 because it is used to access kubernetes API server"
+	}
+	return false, ""
+}
+
+func conflictSSH(rule IngressNodeFirewallProtocolRule) bool {
+	return rule.Protocol == ProtocolTypeTCP && rule.ProtocolRule != nil && rule.ProtocolRule.Port == 22 && rule.Action == IngressNodeFirewallDeny
+}
+
+func conflictDHCP(rule IngressNodeFirewallProtocolRule) bool {
+	return rule.Protocol == ProtocolTypeUDP && rule.ProtocolRule != nil && rule.ProtocolRule.Port == 68 && rule.Action == IngressNodeFirewallDeny
+}
+
+func conflictETCD(rule IngressNodeFirewallProtocolRule) bool {
+	return rule.Protocol == ProtocolTypeTCP && rule.ProtocolRule != nil &&
+		(rule.ProtocolRule.Port == 2380 || rule.ProtocolRule.Port == 2379) && rule.Action == IngressNodeFirewallDeny
+}
+
+func conflictAPIServer(rule IngressNodeFirewallProtocolRule) bool {
+	return rule.Protocol == ProtocolTypeTCP && rule.ProtocolRule != nil && rule.ProtocolRule.Port == 6443 && rule.Action == IngressNodeFirewallDeny
+}
```

Comment:
regardless to the action we need to warn user touching those APIs
can u have a lookup table indexed by protocol to get all restricted ports the see if the configured port or port range will include any of those pinhole ports ? that will make it easy to add ports if needed w/o changing the code

Reply:
Done